### PR TITLE
Update default exec spaces

### PR DESCRIPTION
**Pull request by jeffmiles63**
This PR is a replacement of the #ifdef'd setting of the Default execution spaces with an adaptation of space priorities determined by a selectable configuration.  The SpaceProperty and SpacePriority are specialized in the Fwd_SPACE header files which is consistent with the previous backend refactoring.  There is also a new generated header file which operates similarly to the other refactor backend headers where the Enum which drives the SpaceProperty selection is generated from CMake Enabled Backends.  The only two options are the priority configuration settings for default exec space and default host exec space.

Note that this PR is execution spaces only.  The memory spaces will be in a different PR due to the multitude of files that will be changed.

### CODE REVIEW

File: core/src/Kokkos_Concepts.hpp [hunk 1]
```diff
@@ -339,19 +339,7 @@ struct is_space {
...
-      Kokkos::DefaultHostExecutionSpace, execution_space>::type;
-#else
-#if defined(KOKKOS_ENABLE_OPENMPTARGET)
-  using host_execution_space = typename std::conditional<
-      std::is_same<execution_space, Kokkos::Experimental::OpenMPTarget>::value,
-      Kokkos::DefaultHostExecutionSpace, execution_space>::type;
-#else
-  using host_execution_space = execution_space;
-#endif
-#endif
```
Comment (masterleinad) [inline core/src/Kokkos_Concepts.hpp:354]: What happened here?

File: core/src/fwd/Kokkos_Fwd_CUDA.hpp [hunk 2]
```diff
@@ -61,6 +61,18 @@ void cuda_prefetch_pointer(const ExecSpace& /*space*/, const void* /*ptr*/,
...
+template <typename ConfigT>
+struct SpacePriority<Kokkos::Cuda, ConfigT>
+    : std::integral_constant<int, 50> {};
+
+template <>
+struct SpacePriority<Kokkos::Cuda, Kokkos::HostParallelPriorityConfig>
+    : std::integral_constant<int, 999> {};
+
+template <>
+struct SpaceProperty<CUDA_idx> {
```
Comment (masterleinad) [inline core/src/fwd/Kokkos_Fwd_CUDA.hpp:73]: Where does this come from?

File: core/src/impl/Kokkos_SpacePriorities.hpp [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+#include <tuple>
+#include <utility>
+#include <string>
+#include <iostream>
+#include <typeinfo>
+
+#include <cxxabi.h>
+
+#ifndef __SPACE_PRIORITIES_HPP
+#define __SPACE_PRIORITIES_HPP
```
Comment (masterleinad) [inline core/src/impl/Kokkos_SpacePriorities.hpp:12]: Let's not use double underscores.
```suggestion
#ifndef KOKKOS_SPACE_PRIORITIES_HPP
#define KOKKOS_SPACE_PRIORITIES_HPP
```

**Review by masterleinad — COMMENTED** [review]
It looks like `SYCL` is missing.

File: Makefile.kokkos [hunk 4]
```diff
@@ -1339,8 +1354,7 @@ KOKKOS_OBJ_LINK = $(notdir $(KOKKOS_OBJ))
 include $(KOKKOS_PATH)/Makefile.targets
 
 kokkos-clean:
-	rm -f $(KOKKOS_OBJ_LINK) KokkosCore_config.h KokkosCore_config.tmp libkokkos.a KokkosCore_Config_SetupBackend.hpp \
-	KokkosCore_Config_FwdBackend.hpp KokkosCore_Config_DeclareBackend.hpp
+	rm -f $(KOKKOS_OBJ_LINK) KokkosCore_config.h KokkosCore_config.tmp libkokkos.a KokkosCore_Config_*
```
[the remark this answers is not in the document]
Comment (jeffmiles63) [reply Makefile.kokkos:1357]: There are a large number of KokkosCore_Config_  files generated.

**Comment by jeffmiles63** [comment]
> It looks like `SYCL` is missing.

Added with 2e989f9da4bbf3ffbdee3e0b9c03c4b2df8de3d2

File: Makefile.kokkos (hunk 4 above)
[the remark this answers is not in the document]
Comment (masterleinad) [reply Makefile.kokkos:1357]: OK. Fair enough.

**Comment by masterleinad** [comment]
Looks like this still has conflicts with `develop`.

File: cmake/kokkos_functions.cmake [hunk 5]
```diff
@@ -987,3 +987,23 @@ FUNCTION(KOKKOS_CONFIG_HEADER SRC_FILE TARGET_FILE HEADER_GUARD HEADER_PREFIX DA
...
+##                        DATA_LIST list of backends to include in generated file
+FUNCTION(KOKKOS_CONFIG_SPACE_ENUM SRC_FILE TARGET_FILE HEADER_GUARD ENUM_LEN_TAG DATA_LIST)
+   SET(HEADER_GUARD_TAG "${HEADER_GUARD}_HPP_")
+   CONFIGURE_FILE(cmake/${SRC_FILE} ${PROJECT_BINARY_DIR}/temp/${TARGET_FILE}.work COPYONLY)
+   FOREACH( BACKEND_NAME ${DATA_LIST} )
+     SET(NEXT_SPACE_IDX "${BACKEND_NAME}_idx, \@NEXT_SPACE_IDX\@")
+   CONFIGURE_FILE(${PROJECT_BINARY_DIR}/temp/${TARGET_FILE}.work ${PROJECT_BINARY_DIR}/temp/${TARGET_FILE}.work @ONLY)
+   ENDFOREACH()
+   SET(NEXT_SPACE_IDX "${ENUM_LEN_TAG}" )
+   CONFIGURE_FILE(${PROJECT_BINARY_DIR}/temp/${TARGET_FILE}.work ${TARGET_FILE} @ONLY)
```
Comment (dalg24) [inline cmake/kokkos_functions.cmake:1008]: No.  Do the loop in CMake and achieve this in a single `configure_file` call.

File: cmake/KokkosCore_Config_SpaceEnum.in [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+*/
+#ifndef @HEADER_GUARD_TAG@
+#define @HEADER_GUARD_TAG@
+
+namespace Kokkos {
+namespace Impl {
+
+enum {
+   @NEXT_SPACE_IDX@
+};
```
Comment (dalg24) [inline cmake/KokkosCore_Config_SpaceEnum.in:51]: Is the sole purpose of this enumeration to determine the number of execution spaces?
If so did you consider alternatives such as defining a type such as `using ExecutionSpaces = std::tuple<Foo, Bar, Baz>;`?

We could certainly use something like this downstream in ArborX
https://github.com/arborx/ArborX/blob/f11c397db90221a6b4ea2dcb1403a53cfd8ef7f7/test/CMakeLists.txt#L2-L31

File: core/src/fwd/Kokkos_Fwd_CUDA.hpp (hunk 2 above)
Comment (jrmadsen) [reply core/src/fwd/Kokkos_Fwd_CUDA.hpp:73]: I'll post the original prototype, which might help demonstrate why it's done this way (and allow you to play around with it if you want)

**Comment by jrmadsen** [comment]
Here is a quick little prototype for the changes here that you can experiment with. It treats priorities like `nice`, e.g. lower number are higher priority

## Prototype

```cpp

#include <type_traits>
#include <tuple>
#include <utility>
#include <string>
#include <iostream>
#include <typeinfo>

#include <cxxabi.h>

//------------------------------------------------------------------------------------------//
// various spaces including the user plugin
//------------------------------------------------------------------------------------------//
struct DummySpace {};
struct CudaSpace {};
struct CudaUVMSpace {};
struct OpenMPSpace {};
struct Serial {};
struct UserPlugin {};

//------------------------------------------------------------------------------------------//
// Priority configuration types
//------------------------------------------------------------------------------------------//
struct DefaultPriorityConfig {};
struct CustomPriorityConfig {};

template <typename T, typename ConfigSet = DefaultPriorityConfig>
struct SpacePriority : std::integral_constant<int, 0> {};

//------------------------------------------------------------------------------------------//
// Pre-processor defining which priority configuration to be used
//------------------------------------------------------------------------------------------//
#if !defined(KOKKOS_PRIORITY_SET)
#define KOKKOS_PRIORITY_SET DefaultPriorityConfig
#endif

//------------------------------------------------------------------------------------------//
// Set default priorities
//------------------------------------------------------------------------------------------//
// always lowest
template <typename ConfigT>
struct SpacePriority<DummySpace, ConfigT> : std::integral_constant<int, 1000> {
};

template <typename ConfigT>
struct SpacePriority<Serial, ConfigT> : std::integral_constant<int, 100> {};

template <typename ConfigT>
struct SpacePriority<OpenMPSpace, ConfigT> : std::integral_constant<int, 90> {};

template <typename ConfigT>
struct SpacePriority<CudaSpace, ConfigT> : std::integral_constant<int, 80> {};

template <typename ConfigT>
struct SpacePriority<CudaUVMSpace, ConfigT> : std::integral_constant<int, 70> {
};

//------------------------------------------------------------------------------------------//
// For CustomPriorityConfig, we redefine the CudaUVMSpace to be lower than CudaSpace
//------------------------------------------------------------------------------------------//
template <>
struct SpacePriority<CudaUVMSpace, CustomPriorityConfig>
    : std::integral_constant<int, 85> {};

template <>
struct SpacePriority<UserPlugin> : std::integral_constant<int, -10> {};

enum {
  Serial_idx = 0,
  CudaSpace_idx,
  OpenMPSpace_idx,
  CudaUVMSpace_idx,
  TotalNumberOfNativeSpaces,
  TotalNumberOfSpaces = TotalNumberOfNativeSpaces + 10
};

template <size_t Idx>
struct SpaceProperty {
  using type = Serial;
};

template <typename... T>
struct TypeList;

template <template <typename, typename> class PrioT, typename Curr,
          typename Tail>
struct GetPriority;

template <template <typename, typename> class PrioT, typename Curr>
struct GetPriority<PrioT, Curr, TypeList<>> {
  using type = Curr;
};

template <template <typename, typename> class PrioT, typename Curr, typename Tp,
          typename... Tail>
struct GetPriority<PrioT, Curr, TypeList<Tp, Tail...>> {
  using type = std::conditional_t<
      (PrioT<Curr, KOKKOS_PRIORITY_SET>::value <
       PrioT<Tp, KOKKOS_PRIORITY_SET>::value),
      typename GetPriority<PrioT, Curr, TypeList<Tail...>>::type,
      typename GetPriority<PrioT, Tp, TypeList<Tail...>>::type>;
};

template <size_t... Idx>
struct GetTypesT {
  using type = TypeList<typename SpaceProperty<Idx>::type...>;
};

template <typename Tp>
struct GetTypes;

template <size_t... Idx>
struct GetTypes<std::index_sequence<Idx...>> : GetTypesT<Idx...> {};

inline std::string demangle(const char* _cstr) {
  int _ret      = 0;
  char* _demang = abi::__cxa_demangle(_cstr, nullptr, nullptr, &_ret);
  if (_demang && _ret == 0)
    return std::string(const_cast<const char*>(_demang));
  return _cstr;
}

inline std::string demangle(const std::string& _str) {
  return demangle(_str.c_str());
}

template <typename Tp>
inline std::string demangle() {
  return demangle(typeid(Tp).name());
}

#define SPACE_PROPERTY_IDX(A, B) A##B
#define SET_SPACE_PROPERTY(SPACE)                         \
  template <>                                             \
  struct SpaceProperty<SPACE_PROPERTY_IDX(SPACE, _idx)> { \
    using type = SPACE;                                   \
  };
#define SET_USER_SPACE_PROPERTY(SPACE, IDX)               \
  template <>                                             \
  struct SpaceProperty<TotalNumberOfNativeSpaces + IDX> { \
    using type = SPACE;                                   \
  };

SET_SPACE_PROPERTY(CudaSpace)
SET_SPACE_PROPERTY(CudaUVMSpace)
SET_SPACE_PROPERTY(OpenMPSpace)
SET_SPACE_PROPERTY(Serial)
// SET_USER_SPACE_PROPERTY(UserPlugin, 0)

using DefaultExecutionSpace =
    typename GetPriority<SpacePriority, DummySpace,
                         typename GetTypes<std::make_index_sequence<
                             TotalNumberOfSpaces>>::type>::type;

int main() {
  std::cout << "Default execution space = " << demangle<DefaultExecutionSpace>()
            << std::endl;
  return 0;
}
```

## Example

```console
$ clang++ -std=c++14 priority.cpp -o priority && ./priority
Default execution space = CudaUVMSpace
$ clang++ -DKOKKOS_PRIORITY_SET=DefaultPriorityConfig -std=c++14 priority.cpp -o priority && ./priority
Default execution space = CudaUVMSpace
$ clang++ -DKOKKOS_PRIORITY_SET=CustomPriorityConfig -std=c++14 priority.cpp -o priority && ./priority
Default execution space = CudaSpace
```

File: cmake/KokkosCore_Config_SpaceEnum.in (hunk 6 above)
Comment (masterleinad) [reply cmake/KokkosCore_Config_SpaceEnum.in:51]: This is actually `GetTypes`.

File: core/src/impl/Kokkos_SpacePriorities.hpp [hunk 7]
```diff
@@ -0,0 +1,98 @@
...
+template <size_t... Idx>
+struct GetTypesT {
+  using type = TypeList<typename SpaceProperty<Idx>::type...>;
+};
+
+template <typename Tp>
+struct GetTypes;
+
+template <size_t... Idx>
+struct GetTypes<std::index_sequence<Idx...>> : GetTypesT<Idx...> {};
```
[the remark this answers is not in the document]
Comment (jrmadsen) [reply core/src/impl/Kokkos_SpacePriorities.hpp:64]: Yeah you can. That's probably just a leftover from a previous iteration of the prototype

**Comment by crtrott** [comment]
@nmm0 Do you still need this to get worked on? If not please close.

**Pull request closed without merging**